adserver/tasks.py

```diff
@@ -227,26 +239,36 @@ def daily_update_impressions(day=None):
 
     log.info("Updating AdImpressions for %s-%s", start_date, end_date)
 
-    for impression_type in IMPRESSION_TYPES:
-        queryset = _default_filters(impression_type, start_date, end_date)
-
-        for values in (
-            queryset.values("publisher", "advertisement")
-            # This needs to be publisher and not advertisement to gets decisions properly
-            .annotate(total=Count("publisher"))
-            .filter(total__gt=0)
-            .order_by("-total")
-            .iterator()
-        ):
+    for values in (
+        Offer.objects.using(settings.REPLICA_SLUG)
+        .filter(
+            date__gte=start_date,
+            date__lt=end_date,  # Things at UTC midnight should count towards tomorrow
+        )
+        .values("publisher", "advertisement")
+        # This needs to be publisher and not advertisement to gets decisions properly
+        .annotate(
+            total_decisions=Count("publisher"),
+            total_offers=Count("publisher", filter=Q(advertisement__isnull=False)),
+            total_views=Count("publisher", filter=Q(viewed=True)),
+            total_clicks=Count("publisher", filter=Q(clicked=True)),
```

Comment:
I wonder about some way to pull this query into a helper function? We'd probably need to pass the `values` and `primary_value` (which would be `publisher` in this case). That way it reduces errors around filtering differently. 

All of the queries tend to be slightly different though, so maybe it doesn't make sense. I think there's a reason I ended up with `_default_filters` instead of a larger helper function. 🤷